### PR TITLE
SQL: Create millisecond precision timestamp literals.

### PR DESCRIPTION
**Pull request by gianm**
Fixes a bug where implicit casts of strings to timestamps would use seconds
precision rather than milliseconds. The new test case
testCountStarWithBetweenTimeFilterUsingMillisecondsInStringLiterals
exercises this.

### CODE REVIEW

**Review by clintropolis — APPROVED** [review]
:+1:

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java [hunk 1]
```diff
@@ -82,12 +83,18 @@
...
+      DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS");
 
   private static final Charset DEFAULT_CHARSET = Charset.forName(ConversionUtil.NATIVE_UTF16_CHARSET_NAME);
 
   private static final Pattern TRAILING_ZEROS = Pattern.compile("\\.?0+$");
 
+  /**
+   * Milliseconds precision.
+   */
+  private final static int TIMESTAMP_LITERAL_PRECISION = 3;
```
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:96]:
```suggestion
  private static final int TIMESTAMP_LITERAL_PRECISION = 3;
```

This fixes the checkstyle checking.

Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java]: Thanks!

**Comment by FrankChen021** [comment]
@gianm There're some failed cases and looks like they're related to the changes in this PR.

```bash
[ERROR] org.apache.druid.sql.calcite.CalciteSelectQueryTest.testSelectCurrentTimeAndDateLosAngeles  Time elapsed: 0.064 s  <<< FAILURE!
java.lang.AssertionError: 
Conversion to relational algebra failed to preserve datatypes:
validated type:
RecordType(TIMESTAMP(0) NOT NULL CURRENT_TIMESTAMP, DATE NOT NULL CURRENT_DATE, DATE NOT NULL EXPR$2) NOT NULL
converted type:
RecordType(TIMESTAMP(3) NOT NULL CURRENT_TIMESTAMP, DATE NOT NULL CURRENT_DATE, DATE NOT NULL EXPR$2) NOT NULL
rel:
LogicalProject(CURRENT_TIMESTAMP=[1999-12-31 16:00:00:TIMESTAMP(3)], CURRENT_DATE=[1999-12-31], EXPR$2=[+(1999-12-31, 86400000:INTERVAL DAY)])
  LogicalValues(tuples=[[{ 0 }]])
	at org.apache.calcite.sql2rel.SqlToRelConverter.checkConvertedType(SqlToRelConverter.java:458)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:575)
	at org.apache.calcite.prepare.PlannerImpl.rel(PlannerImpl.java:254)
```

**Comment by gianm** [comment]
> @gianm There're some failed cases and looks like they're related to the changes in this PR.

Ah, thanks. I looked into these, and there were some places I missed when setting the default precision to millis. I think the latest patch gets them all. I also added some additional tests to make sure.

**Review by FrankChen021 — APPROVED** [review]

**Pull request merged**